### PR TITLE
[LTX-2.5] Refactor LTX-2.5 Diffusion Decoder Forward Methods

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR refactors the `LTX2VideoDiffusionDecoder3d` `forward` method as follows:

- `forward` represents a single Stage 5 diffusion denoising step, following other DiTs in `/models/transformers`
- `forward_stages_1_to_3` and `forward_stage_4` have been renamed to `encode_context_stages_1_to_3` and `encode_context_stage_4`, respectively, since they calculate the conditioning for the Stage 5 denoising loop using the input `latents` from the main LTX-2.5 DiT.
- The denoising loop is now performed by `_denoise` in the `ModelMixin` subclass `LTX2VideoDiffusionDecoderModel` rather than `LTX2VideoDiffusionDecoder3d`.

The motivation for the refactor is to have a single `forward` method while retaining the VAE tiling interface for `LTX2VideoDiffusionDecoderModel`. See https://github.com/huggingface/diffusers/pull/14447#discussion_r3759775444 for more info.

## Before submitting
- [x] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR? If so:
  - [x] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [x] Did you run the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on the diff?
  - [x] Did you share the final self-review notes in the PR description or a comment?
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [x] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)? (important for complex PRs)
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?
- [ ] Are you the author (or part of the team) of the model/pipeline (only applicable for model/pipeline related PRs)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu
@sayakpaul

### CODE REVIEW

**Comment by dg845** [comment]
<details open>
    <summary>Self-Review Report</summary>

## Self-review

Ran the project's self-review rubric (`.ai/references/review-rules.md` plus `code_style.md`, `models.md`, `testing.md`,
`pitfalls.md`) over this diff.

**Scope:** `src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py` (+99 −78),
`tests/models/autoencoders/test_models_ltx2_diffusion_decoder.py` (+51).

**Verdict: READY** — no blocking issues. Three items below are deliberate choices I'd rather surface than leave silent, and one is
a design question for the reviewer.

---

### Blocking issues

None.

### Non-blocking / for discussion

**1. `_denoise` is a single-caller private helper.**
`ltx2_diffusion_decoder.py:829`, called only from `:986`. `code_style.md` says *"If a private helper has only one caller, inlining
it at the call site is usually the cleaner choice."* I kept it separate because inlining puts a 20-line Euler loop inside an
already three-deep tile loop. Flagging because the rule points the other way — happy to inline if you prefer.

**2. `_decode(z, tiled=...)` has the shape of a config flag.**
`ltx2_diffusion_decoder.py:878`. It isn't one: two real callers pass different values (`tiled_decode` → `True`, `decode` →
`self.use_tiling`). Noting it because the shape is what `code_style.md`'s "no configuration options" rule targets.

**3. The module's full computation no longer lives in any single `forward`. (Design question.)**
`models.md` says *"All layer calls should be visible directly in `forward`."* The new `forward` satisfies that for stage 5 — an
improvement on the old one, which hid two whole stages behind `forward_stage_4(forward_stages_1_to_3(x))`. But
`LTX2VideoDiffusionDecoder3d.forward` no longer decodes; it is one denoising step, and stages 1–4 are reached as named context methods. That is the point of the PR, and it is the thing most worth a second opinion.

**4. One intentional numerical change.**
Multi-step x0 now returns the final x0 prediction directly instead of running an Euler update to `t=0`. The update reduces to that prediction (`x_t − t·(x_t − pred)/t = pred`), so this is the same value without a full-canvas float32 round trip, and it matches the reference decoder's `_decode_one_tile`. Measured `~2e-07` in float32; every other path is bit-exact.

### Addressed while self-reviewing

- **Test gap.** `tiled_decode` is public and must tile regardless of `use_tiling` (which gates only `decode`'s routing) — an
intermediate version of this refactor broke that, and the whole suite stayed green, because a fallback to one full-grid tile
reproduces the untiled decode exactly. Added `test_tiled_decode_tiles_even_when_tiling_is_disabled`, which counts stage-4
invocations instead of comparing outputs. Verified it fails on the broken version while the other three tiling tests still pass.
- **Dead code.** Removed `self.spatial_compression_ratio` / `self.temporal_compression_ratio` from the model — this diff removed their last reader. Non-breaking: `ConfigMixin.__getattr__` still resolves the attribute to the config value with the standard deprecation warning. The inner decoder's own `temporal_compression_ratio` (`:526`, read at `:620`) stays; a plain `nn.Module` has no such fallback.

### Verification

- **Numeric baseline vs. the pre-refactor code** (tiny config, 6 configurations): bit-exact for single-step x0 and for the velocity
path on both tiled and untiled; `~2e-07` for the two multi-step x0 cases, which is item 4 above.
- **Real LTX-2.5 checkpoint at 768×512×121:** stages 1–3 → `(1, 69, 64, 96, 512)`, stage 4 context → `(1, 121, 128, 192, 256)`, output `(1, 3, 121, 512, 768)`, finite, on both decode paths. Also confirmed `use_tiling=False` runs stage 4 exactly once on the
full grid while `enable_tiling()` runs it twice — i.e. the routing genuinely honors the flag at production dimensions.
- **Accelerate hooks.** The untiled path changed from `self.decoder(z, ...)` (an `nn.Module.__call__`) to a bound method, so a hook on `self.decoder` no longer fires for it. Checked because it would break silently: not a regression, since the tiled path never invoked `decoder.__call__` either and the model already relies on submodule-level hooks. Both paths are now consistent. Offload/parallelism tests pass.
- **Suite:** 45 passed, 5 skipped. `ruff check`, `ruff format`, `make fix-copies` clean; the `# Copied from` blend helpers are
untouched and in sync.

### Dead code analysis

| symbol | verdict | reason |
| --- | --- | --- |
| `_denoise` (`:829`) | Used | `:986` |
| `_decode` (`:878`) | Used | `:876` (`tiled_decode`), `:1032` (`decode`) |
| `tiled_decode` (`:860`) | Used | public API + test `:172` |
| `encode_context_stages_1_to_3` (`:580`), `encode_context_stage_4` (`:603`) | Used | `:912`, `:961` |
| inner `forward` (`:623`) | Used | `:844` |
| inner `out_channels`, `default_num_inference_steps`, `model_output_type`, `trailing_pad_latent_frames`, `patch_size`,
`context_channels` | Used | all still read from `_decode` / `_denoise` / `__init__` — re-checked after deleting the old inner
`forward`, which had been their reader |

Nothing newly orphaned.

### Docs

No staleness found. `docs/source/en/api/models/ltx2_diffusion_decoder.md` documents `enable_tiling()` semantics and the
remnant-merge rule, both unchanged; the renamed methods were never in public docs.

Two rules this PR surfaced that aren't written down anywhere, offered for the agent guides:

1. *When a model exposes both `decode` and `tiled_decode`, `tiled_decode` must tile unconditionally — `use_tiling` gates only whether `decode` routes to it.* Every in-tree autoencoder follows this implicitly; nothing states it, and it cost a real bug here.
2. *Removing an attribute that mirrors a `register_to_config` value is non-breaking (`ConfigMixin.__getattr__` back-stops it) — but only when it really is a mirror.* On `AutoencoderKLLTX2Video` the same-named attributes are **derived** (config defaults to `None`, value computed from the block config, read internally at `:1274-1276` / `:1364`), so the same removal would break it. The two files sit side by side and the pattern does not transfer.

</details>

File: src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py [hunk 1]
```diff
@@ -476,10 +476,14 @@ def forward(self, hidden_states: torch.Tensor, drop_leading_frame: bool = True)
...
-    `model_output_type="x0"` and a single step — how LTX-2.5 ships — stage 5 runs once and its prediction *is* the
-    output; more steps add reverse Euler updates.
+    Stages 1-4 deterministically upsample the latent into a context volume with neighborhood-attention blocks; that
+    volume conditions stage 5, which is an ordinary diffusion transformer over patchified pixels. [`forward`] is one
+    stage-5 step, so this module is the network alone — the timestep schedule, the noise and the reverse Euler updates
+    live on [`LTX2VideoDiffusionDecoderModel`], which drives it.
+
+    The context is not cross-attended the way a prompt embedding would be: it is projected into the residual stream of
+    every block and shares the token grid with the pixels, which is what lets a tile of context stand in for a tile of
+    output during tiled decoding.
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py]: I am not sure we need all the details on what is cross-attended and what is not. I believe we could simplify it like so:


```suggestion
    Stages 1-4 deterministically upsample the input latent into a context volume with neighborhood-attention blocks; that
    volume conditions stage 5.
```

File: src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py [hunk 2]
```diff
@@ -952,7 +983,7 @@ def tiled_decode(
                             h0 * scale_h : h0 * scale_h + tile_pixel_shape[3],
                             w0 * scale_w : w0 * scale_w + tile_pixel_shape[4],
                         ]
-                    row.append(decoder.denoise(context, x_t, num_inference_steps))
+                    row.append(self._denoise(context, x_t, num_inference_steps))
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py:957]: I am okay to keep `_denoise()` as is but since it's not shared anywhere maybe we can fold that in here?

File: src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py [hunk 3]
```diff
@@ -998,17 +1029,7 @@ def decode(
...
-        tile_latent_min_num_frames = self.tile_sample_min_num_frames // self.temporal_compression_ratio
-        if self.use_tiling and (
-            z.shape[2] > tile_latent_min_num_frames
-            or z.shape[3] > tile_latent_min_height
-            or z.shape[4] > tile_latent_min_width
-        ):
-            decoded = self.tiled_decode(z, generator=generator, num_inference_steps=num_inference_steps)
-        else:
-            decoded = self.decoder(z, generator=generator, num_inference_steps=num_inference_steps)
+        decoded = self._decode(z, tiled=self.use_tiling, generator=generator, num_inference_steps=num_inference_steps)
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py]: But we follow a clear conditional pattern in others no?

(Flux2 autoencoder follows this):

```py
    def _decode(self, z: torch.Tensor, return_dict: bool = True) -> DecoderOutput | torch.Tensor:
        if self.use_tiling and (z.shape[-1] > self.tile_latent_min_size or z.shape[-2] > self.tile_latent_min_size):
            return self.tiled_decode(z, return_dict=return_dict)

        if self.post_quant_conv is not None:
            z = self.post_quant_conv(z)

        dec = self.decoder(z)
```

File: src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py [hunk 4]
```diff
@@ -852,6 +826,37 @@ def blend_t(self, a: torch.Tensor, b: torch.Tensor, blend_extent: int) -> torch.
             )
         return b
 
+    def _denoise(self, latent_context: torch.Tensor, x_t: torch.Tensor, num_inference_steps: int) -> torch.Tensor:
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py:823]: But I guess the denoise logic should go to the pipeline no? Cc:  @yiyixuxu

File: src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py (hunk 2 above)
Comment (dg845) [reply src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py:957]: My motivation for having `_denoise` as a separate method is so that the code is easier to follow (since `_denoise` is called inside a nested loop over `temporal_tiles`/`height_tiles`/`width_tiles`, inlining the denoising loop would make `_decode` even more complicated).

File: src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py (hunk 4 above)
Comment (dg845) [reply src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py:823]: The current PR implementation (with denoising in `decode`) has the advantage that it doesn't require changes to the current `LTX2VideoDiffusionDecodePipeline` standard pipeline or the `LTX2DiffusionVaeDecoderStep` modular block, but I think it's reasonable if we want to refactor those as well to fit the diffusion pipeline design better.

Comment (sayakpaul) [reply src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py:823]: I think that is more consistent with how we do it for other pipelines. But let's see what @yiyixuxu has to say about this.

Comment (yiyixuxu) [reply src/diffusers/models/autoencoders/ltx2_diffusion_decoder.py:823]: it should go to the pipeline

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_diffusion_decode.py [hunk 5]
```diff
@@ -24,6 +28,245 @@
...
+            model_output = (x_t.float() - model_output) / sigma
+        # `step` returns in `model_output`'s dtype, float32 above to keep the update off the canvas dtype; the
+        # canvas itself stays in the dtype its noise was drawn in.
+        x_t = scheduler.step(model_output, t, x_t, return_dict=False)[0].to(dtype)
+        if progress_bar is not None:
+            progress_bar.update()
+    return x_t
+
+
+def _tiled_decode(
```
Comment (dg845) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_diffusion_decode.py:137]: `_tiled_decode` and `_denoise` are currently module-level functions rather than being in `__call__`; the motivation is so that the tiling and denoising logic can be reused by the modular pipeline (`LTX2DiffusionVaeDecoderStep`) via `# Copied from`. See the thread beginning at https://github.com/huggingface/diffusers/pull/14694#discussion_r3923243431 for more context.

File: src/diffusers/modular_pipelines/ltx2/decoders.py [hunk 6]
```diff
@@ -197,8 +477,13 @@ def __call__(self, components, state: PipelineState) -> PipelineState:
...
         latents = latents.to(decoder.dtype)
-        # It samples the noise it denoises, so pass the generator to keep decoding reproducible.
-        video = decoder.decode(latents, generator=block_state.generator, return_dict=False)[0]
+        # The decoder's `forward` is a single denoising step, so the loop (and the tiling around it) is driven
+        # from here. It samples the noise it denoises, so pass the generator to keep decoding reproducible.
+        scheduler = components.diffusion_decoder_scheduler
+        _check_scheduler(scheduler)
+        sigmas = _decoder_sigmas(decoder, block_state.decode_num_inference_steps)
+        decode = _tiled_decode if _should_tile(decoder, latents) else _untiled_decode
+        video = decode(decoder, scheduler, latents, block_state.generator, sigmas)
```
Comment (dg845) [inline src/diffusers/modular_pipelines/ltx2/decoders.py:486]: Currently the modular diffusion decoder denoising loop is implemented through `_tiled_decode` (which calls `_denoise` internally) rather than through a loop modular block like `LTX2DenoiseLoopWrapper`. The motivation is so that `LTX2DiffusionVaeDecoderStep` remains interchangeable with `LTX2VaeDecoderStep`. (In general the diffusion decoder performs an outer tiling loop over `t`/`h`/`w` tiles and an inner denoising loop on each tile, so it also needs nested loop support.) See also https://github.com/huggingface/diffusers/pull/14694#discussion_r3974730649.
